### DISCUSSION TITLE
Exporting and Importing complete Dashborads

### DISCUSSION THREAD

**Discussion by fbgeobit** [apache/superset#30955] [Q&A / Help]
At our organization we use a develop and production Superset structure, so we have at least:
- A Superset for development
- A Superset for production
We think Superset is a powerful tool, but at same time, we also found too many problems trying to sustain this use case. Seems it's not being developed for admin usage:
- Does not exists yet a secure way to export and import (or reimport) properly finished dashboards, without a lot of post-work after each import (taking into account we are using same Superset version on both environments).
- There exists a way to bulk export/import roles and users?
- There is not a way for filtering datasets based on what dashboards there are being used, so there is not a way to know what are the interdependencies between dashboards-charts-datasets. This is a key matter when you need to move your current 10's of dashboards to a new version without breaking anything.
- We currently found what we think it's a bug:
    - If your import an existing dashboard to overwrite current one because you want to upgrade it, only the dashboards itself is overwritten, but charts or datasets remain intact, braking our develop-production model.
 
Please, take this post as a constructive critic (we really appreciate you work): if the software is not production ready or administration teams need to put hours and hours to configure dashboards that worked previously on development environment to provide a simple upgrade, then is not very comfortable using it.

Have you any recommendation for get the best using Superset in a development --> production setting? I read some posts about this, but with no answers yet (https://github.com/apache/superset/discussions/30288, https://github.com/apache/superset/discussions/28693).

Thanks in advance.

**Comment by mistersister** [comment]
Agreed. Bulk export/import is not working as expected. 

After using superset export-dashboard tool to backup everything on a 4.0.2 install, trying to import on another 4.0.2 install was met with errors and had to manually fix the exported files one by one before I could import.

Maybe I did something wrong but the import/export mechanisms could use a second look.

**Reply by fbgeobit** [reply]
Right. 
An updated documentation about using options with export-dashboard tool would help...

**Reply by johna-xceleration** [reply]
@mistersister - when you say "was met with errors", do you mean during the import or with the imported dashboard? Also, can you provide a rough idea of what "manually fix" means?

I'm able to export from one instance and import into the other. The import succeeds however the dashboard fails to load any of the charts (even though the charts themselves are fine).

**Reply by mistersister** [reply]
@johna-xceleration I don't recall exactly but I had to fix the exported files manually by editing the yaml files.

There was a field called `template_params=''` that apparently needed to `template_params=null` in order to go through the webapp import wizard.

Might be that the dashboard was made and exported on a very old version of Superset or maybe something went wrong during the export process but anyhow these days we backup up the whole database volume (using docker) and that seems to work everytime for me.

**Comment by seanfulton** [comment]
Any further on this? We are seeing a number of issues in 5.0.0 with export -> import. Seems to be data is missing on the export, even though there are no errors.

[stats_logger] (incr) DashboardRestApi.export.success
[stats_logger] (timing) DashboardRestApi.export.time | 3219.9838403612375

**Comment by scpcy** [comment]
I have the same problem trying to import dashboards from staging to production environment.

**Comment by fbgeobit** [comment]
To solve the issues we encountered when importing dashboards from the staging environment into production, our team developed the following operational procedure. It works reasonably well for us, although it is somewhat tedious.

**A) In the staging or development environment:**

1. Assign a unique ID to the dashboard name in order to identify that specific dashboard unambiguously.
2. Propagate that dashboard ID to the names of all related charts.
3. Propagate that dashboard ID to the names of all related datasets.
4. Export complete dashboard.

This allows us to identify every dataset and chart associated with a specific dashboard.


**B) In the production environment, to avoid undesirable import behaviour:**

1. Search for all datasets associated with the dashboard being updated by using the custom ID, and delete them.
2. Search for all charts associated with the dashboard being updated by using the custom ID, and delete them.
3. Import the updated dashboard together with all its datasets and charts and overwrite it when asked to preserve permissions, embebdings, etc.

Hope that helps you.
